core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -969,435 +915,537 @@ private void syncLocalToRemote() {
+            docConfig
+        ));
+
+        return writeContainer;
+      case APPLY_FROM_REMOTE:
+        final BsonDocument remoteVersion =
+            DocumentVersionInfo.getDocumentVersionDoc(remoteChangeEvent.getFullDocument());
+        return replaceOrUpsertOneFromRemote(nsConfig, docConfig.getDocumentId(),
+            remoteChangeEvent.getFullDocument(), remoteVersion);
+      case CONFLICT:
+        return resolveConflict(nsConfig, docConfig, remoteChangeEvent);
+      case REMOTE_FIND:
+        return remoteFind(nsConfig, docConfig, caller);
+      case DROP_EVENT_AND_DESYNC:
+        return emitErrorAndDesync(nsConfig, docConfig, formattedMessage, ex);
+      case DROP_EVENT_AND_PAUSE:
+        return emitErrorAndPause(docConfig, formattedMessage, ex);
+      case DELETE_LOCAL_DOC:
+        return deleteOneFromRemote(nsConfig, docConfig.getDocumentId());
+      case DELETE_LOCAL_DOC_AND_DESYNC:
+        return desyncDocumentsFromRemote(nsConfig, docConfig.getDocumentId())
+            .withPostCommit(() -> triggerListeningToNamespace(nsConfig.getNamespace()));
+      default:
+        throw new IllegalStateException("unhandled synchronization action: " + action);
+    }
+  }
+
+  private LocalSyncWriteModelContainer remoteFind(
+      final NamespaceSynchronizationConfig nsConfig,
+      final CoreDocumentSynchronizationConfig docConfig,
+      final String caller) {
+    final DocumentVersionInfo lastSeenVersionInfo =
+        DocumentVersionInfo.getLocalVersionInfo(docConfig);
+
+    SyncAction action = null;
+    SyncMessage message = null;
+    ChangeEvent<BsonDocument> remoteChangeEvent = null;
+
+    // fetch the latest version to guard against stale events from other clients
+    final BsonDocument newestRemoteDocument = this.getRemoteCollection(nsConfig.getNamespace())
+        .find(new BsonDocument("_id", docConfig.getDocumentId())).first();
+
+    final boolean isPendingWrite = docConfig.getLastUncommittedChangeEvent() != null;
+
+    if (newestRemoteDocument == null) {
+      // document was deleted remotely
+      if (isPendingWrite) {
+        action = SyncAction.CONFLICT;
+        remoteChangeEvent = ChangeEvents.changeEventForLocalDelete(
+            nsConfig.getNamespace(),
+            docConfig.getDocumentId(),
+            docConfig.hasUncommittedWrites());
+      } else {
+        action = SyncAction.DELETE_LOCAL_DOC_AND_DESYNC;
+      }
+      message = SyncMessage.REMOTE_FIND_DELETED_DOC_MESSAGE;
+    } else {
+      DocumentVersionInfo newestRemoteVersionInfo = null;
+      try {
+        newestRemoteVersionInfo = DocumentVersionInfo.getRemoteVersionInfo(newestRemoteDocument);
+      } catch (final Exception e) {
+        action = SyncAction.DROP_EVENT_AND_DESYNC;
+        message = SyncMessage.CANNOT_PARSE_REMOTE_VERSION_MESSAGE;
+      }
+
+      if (newestRemoteVersionInfo != null) { // if no error has occurred
+        final boolean newestRemoteHasNoVersion = !newestRemoteVersionInfo.hasVersion();
```

Comment:
[nit] in this particular block of code, I think `newestRemoteHasVersion` and `lastSeenHasVersion` make more sense. 

This later on is a mouthful:
```
if (!newestRemoteHasNoVersion && !lastSeenHasNoVersion
```

Reply:
agree to disagree for the sake of getting this done, feel free to refactor on next pass